### ISSUE TITLE
Suggestion : use all fields from source to set destination

### ISSUE THREAD

**Issue by mmaryo** [remondis-it/remap#49]
Hello 

I have a field that is a generated message from some fields in source object

I would like to do something like this 

            Mapping.from(ActivityEntity.class)
                    .to(ActivityDto.class)
                    .set(ActivityDto::getMessage).with(this::getMessage) 
                    .omitInSource(ActivityEntity::getId)
                    .mapper();

    private String getMessage(ActivityEntity entity) {
        switch (entity.getAction()) {
            case ORDER_SHOP_ADD:
                return "Supplier added " + entity.getShop();
            case ORDER_ATTRIBUTED:
                return "Attributed to " + entity.getAttributeTo();
            case ORDER_CLONED:
                return entity.getClone() == null ? "" : entity.getClone().getFirst() + " clone to " + entity.getClone().getSecond();
        }
    }

**Comment by schuettec** [comment]
You can do this with ReMap out of the box. You can make `private String getMessage(ActivityEntity entity)` part of `ActivityEntity` as a read-only getter without parameters. Then you can specify an easy mapping:

```
public class MapperTest {

  public enum Action {
    ORDER_SHOP_ADD,
    ORDER_ATTRIBUTED,
    ORDER_CLONED;
  }

  public static class ActivityEntity {
    private Action action;

    public ActivityEntity() {
      super();
    }

    public ActivityEntity(Action action) {
      super();
      this.action = action;
    }

    public Action getAction() {
      return action;
    }

    public void setAction(Action action) {
      this.action = action;
    }

    public String getMessage() {
      switch (this.getAction()) {
        case ORDER_SHOP_ADD:
          return "ORDER_SHOP_ADD";
        case ORDER_ATTRIBUTED:
          return "ORDER_ATTRIBUTED";
        case ORDER_CLONED:
          return "ORDER_CLONED";
        default:
          return "UNKNOWN";
      }
    }
  }

  public static class ActivityDto {
    private String message;

    public ActivityDto() {
      super();
    }

    public ActivityDto(String message) {
      super();
      this.message = message;
    }

    public String getMessage() {
      return message;
    }

    public void setMessage(String message) {
      this.message = message;
    }
  }

  @Test
  public void multiMapping() {
    Mapper<ActivityEntity, ActivityDto> mapper = Mapping.from(ActivityEntity.class)
        .to(ActivityDto.class)
        .omitInSource(ActivityEntity::getAction)
        .mapper();
    ActivityEntity activityEntity = new ActivityEntity(Action.ORDER_CLONED);
    ActivityDto dto = mapper.map(activityEntity);
    assertEquals("ORDER_CLONED", dto.getMessage());
  }

}
```

**Comment by mmaryo** [comment]
so I need to move display code inside ActivityEntity
ok

**Comment by schuettec** [comment]
There may be another solution if you really need to avoid that, but it seems that `getMessage` needs many fields/methods of the ActivityEntity and therefore it is not that bad to place it there - but this is just my two cents ;-)

Is your problem solved and may I close this one?

**Comment by mmaryo** [comment]
It was a suggestion to get your idea about that feature 
And not sure if you are agree if I develop it

**Comment by schuettec** [comment]
I wonder how ReMap could support testing this.

**Comment by schuettec** [comment]
Hm I think testability is not the problem since we would still check the source/target for fields being mapped. The only thing to note is, that although the provided function in `with(...)` reads a source field, an omit-operation is required in the mapping if the field is not mapped otherwise.

**Comment by mmaryo** [comment]
thanks
maybe I'll write it :)

**Comment by schuettec** [comment]
Yesterday I thought about your suggestion. It would be nice to have this set-operation. I have usecases where I need to set a value that is either calculated from multiple fields of the DTO or simply a static value.
I would suggest the following API:

`Mapping.set(Destination::getField).with(this::valueFunction)` where `valueFunction` can have the following sigantures: 

- `public FieldType valueFunction(Destination d)`
- `public FieldType valueFunction()`

Do you have any plans for contribution? 😉

**Comment by schuettec** [comment]
FYI:  #55

**Comment by schuettec** [comment]
This feature will be available in 4.0.1. Release is triggered.
